rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+* Based on the marks, ensure that the output cannot be used in other contexts than the ephemeral ones (see the User Documentation section for more details on where this is allowed).
+
+> [!TIP]
+>
+> For an example on how to properly introduce a new mark in the outputs, you can check the [PR](https://github.com/opentofu/opentofu/pull/2633) for the deprecated outputs.
+
+Strict rules:
+* A root module cannot define ephemeral outputs since are useless.
+* Any output that wants to use an ephemeral value, it needs to be marked as ephemeral too. Otherwise, it needs to show an error:
+   ```hcl
+    │ Error: Output not marked as ephemeral
+    │
+    │   on mod/main.tf line 33, in output "password":
+    │   33:   value = reference.to.ephemeral.value
+    │
+    │ In order to allow this output to store ephemeral values add `ephemeral = true` attribute to it.
+   ```
+* Any output from a root module that is referencing a write only attribute needs to be marked as sensitive too. Otherwise, an error should be raised
+  ```hcl
+    │ Error: An output referencing a sensitive value needs to be marked with sensitive too
+    │
+    │   on main.tf line 32:
+    │   32: output "write_only_out" {
+    │
+    │ For security reasons, OpenTofu requires any output that is referencing a sensitive value to also be configured the same. If the root module really wants to export this sensitive value, you need to annotate it with the following argument:
+    │     sensitive = true
+  ```
+
+Considering the rules above, root modules cannot have any ephemeral outputs defined.
+
+### Locals
+Any `local` declaration will be marked as ephemeral if in the expression that initialises it an ephemeral value is used:
+```hcl
+variable "var1" {
+  type = string
+}
+
+variable "var2" {
+  type = string
+}
+
+variable "var3" {
+  type      = string
+  ephemeral = true
+}
+
+locals {
+  eg1 = var.var1 == "" ? var.var2 : var.var1 // not ephemeral
+  eg2 = var.var2 // not ephemeral
+  eg3 = var.var3 == "" ? var.var2 : var.var1 // ephemeral because of var3 conditional
+  eg4 = var.var1 == "" ? var.var2 : var.var3 // ephemeral because of var3 usage
+  eg5 = "${var.var3}-${var.var1}" // ephemeral because of var3 usage
+  eg6 = local.eg4 // ephemeral because of eg4 is ephemeral
+}
+```
+
+Once a local is marked as ephemeral, this can be used only in other ephemeral contexts. Check the `User Documentation` section for more details on the allowed contexts.
+
+### Ephemeral resources
+Due to the fact ephemeral resources are not stored in the state/plan file, this block is not creating a diff in the OpenTofu's UI.
+Instead, OpenTofu will notify the user of opening/renewing/closing an ephemeral resource with messages similar to the following:
```

Comment:
Can we discuss what these actions mean? `opening` `renewing` and `closing`?  As someone approaching this fresh I think I can assume what it means but I think we should declare it explicitly so we can have a conversation about it. 

It feels like a strange set of verbs to show up in the ui to me (Opening.. Closing... etc)

Reply:
Changed a little in 24b57e7c7b16816ae8e5d98b57658781656179c9.
But please check the specific chapter dedicated to the `Renew` [method](https://github.com/opentofu/opentofu/blob/6b81cb76c65cb76311db9ecf9d67a94fbb0185cc/rfc/20250317-ephemeral-resources.md#renew-method-details).